File: Diameter of Binary Tree.py
def diameter(root):
    res = 0
    
    def height(node):
        nonlocal res
        if not node:
            return 0
        
        lh = height(node.left)
        rh = height(node.right)
        
        res = max(res, lh + rh)
        
        return max(lh, rh) + 1
    
    height(root)
    return res

File: Diameter of Binary Tree.c
int diameterUtil(struct Node* root, int* height) {
    if (root == NULL) {
        *height = 0;
        return 0;
    }

    int lh = 0, rh = 0;

    int ld = diameterUtil(root->left, &lh);
    int rd = diameterUtil(root->right, &rh);

    *height = (lh > rh ? lh : rh) + 1;

    int curr = lh + rh + 1;

    return (curr > ld && curr > rd) ? curr : (ld > rd ? ld : rd);
}

int diameter(struct Node* root) {
    int height = 0;
    return diameterUtil(root, &height);
}
